### PR TITLE
Fix high cpu sasl loop after disconnect

### PR DESCRIPTION
**Pull request by stefanseufert**
Should fix https://github.com/edenhill/librdkafka/issues/1937

### CODE REVIEW

File: src/rdkafka_transport.c [hunk 1]
```diff
@@ -1403,6 +1403,15 @@ static void rd_kafka_transport_io_event (rd_kafka_transport_t *rktrans,
 
 	case RD_KAFKA_BROKER_STATE_AUTH:
 		/* SASL handshake */
+
+        if (events & POLLHUP) {
```
Comment (edenhill) [inline src/rdkafka_transport.c]: We must read remaining socket buffer data if events&POLLIN, to make sure we receive any auth errors sent by the broker before it disconnects the socket.
So move this check down below the sasl_io_event() call.

File: src/rdkafka_transport.c [hunk 2]
```diff
@@ -1403,6 +1403,15 @@ static void rd_kafka_transport_io_event (rd_kafka_transport_t *rktrans,
...
+            rd_kafka_broker_fail(rkb, LOG_ERR,
```
Comment (edenhill) [inline src/rdkafka_transport.c]: set `errno = EINVAL;` prior to calling broker_fail(). It logs the errno in debug messages and we want to have it set to a known value.

File: src/rdkafka_transport.c (hunk 1 above)
Comment (edenhill) [inline src/rdkafka_transport.c]: Set tab indent to 8 whitespaces.

Comment (stefanseufert) [reply src/rdkafka_transport.c]: Done as requested.

File: src/rdkafka_transport.c (hunk 2 above)
Comment (stefanseufert) [reply src/rdkafka_transport.c]: Done as requested.

File: src/rdkafka_transport.c (hunk 1 above)
Comment (stefanseufert) [reply src/rdkafka_transport.c]: Done as requested.

**Comment by stefanseufert** [comment]
Rebased and squashed fixes into original commit as per contributing guidelines

**Comment by stefanseufert** [comment]
@edenhill anything else you want me to so here? According to your comments in various threads, 11.6 seems to be close and   I really hope to see this included.

**Comment by edenhill** [comment]
0.11.6 would be a stretch, it is way past its deadline and needs to get out the door.
But we're doing 1.0.0 release candidates very soon, which would be a better fit for this.

**Comment by stefanseufert** [comment]
This is a really nasty bug for everyone using SASL auth. A restart of any broker node will immediatedly make the clients go 100% CPU per thread. I've seen boxes logging CPU loads in excess of 350 due to this issue, mailing them completely unaccessible. Since this also affects producers it will result in message loss. There are no known workarounds and there ist no way to detect the problem in the calling code. As far as I can see all platforms are affected. Not sure how much worse a bug could be?

The only reason, why this hasn't cought much more attention seems to be the fact that auth isn't used that commonly.

The fix on the other hand seems trivial ans will only affect those people using auth.

... but it certainly is your call.

**Comment by edenhill** [comment]
Thanks alot for this!
(yes, it will go into 0.11.6 ;) )

**Pull request merged**